### PR TITLE
basic invitelink annotation

### PR DESCRIPTION
**Pull request by heronhaye**

### CODE REVIEW

File: go/client/team_cli_rendering.go [hunk 1]
```diff
@@ -76,44 +79,42 @@ func (c *teamMembersRenderer) outputRole(team, role string, members []keybase1.T
 	}
 }
 
-func (c *teamMembersRenderer) formatInviteName(invite keybase1.AnnotatedTeamInvite) (res string) {
```
Comment (heronhaye) [inline go/client/team_cli_rendering.go]: service formats this as DisplayName now

File: go/teams/list.go [hunk 2]
```diff
@@ -469,85 +471,53 @@ func AnnotateSeitanInvite(ctx context.Context, team *Team, invite keybase1.TeamI
...
+	pkey, err := SeitanDecodePKey(string(invite.Name))
+	if err != nil {
+		return name, err
+	}
+	keyAndLabel, err := pkey.DecryptKeyAndLabel(mctx.Ctx(), team)
+	if err != nil {
+		return name, err
+	}
 
-func AnnotateInvites(ctx context.Context, g *libkb.GlobalContext, team *Team) (AnnotatedInviteMap, error) {
```
Comment (heronhaye) [inline go/teams/list.go]: I deleted this function since it was overkill for the one place that was using it.

File: go/teams/service_helper.go [hunk 3]
```diff
@@ -45,7 +45,7 @@ func GetAnnotatedTeam(ctx context.Context, g *libkb.GlobalContext, id keybase1.T
 	if err != nil {
 		return res, err
 	}
-	det, err := details(ctx, g, t, tracer)
+	det, err := details(mctx, t, tracer)
```
Comment (heronhaye) [inline go/teams/service_helper.go]: Started on this upgrade, probably wasn't worth it in the end

File: go/teams/teams.go [hunk 4]
```diff
@@ -379,18 +379,8 @@ func (t *Team) implicitTeamDisplayName(ctx context.Context, skipConflicts bool)
 	}
 
 	// Add the invites
```
Comment (heronhaye) [inline go/teams/teams.go]: This is the function that was previously using AnnotateInvites( to get usernames for keybase-type invites. Now, it just calls UPAKLoader.LookupUsername directly. Behavior should be the same.

File: go/teams/teams.go [hunk 5]
```diff
@@ -412,17 +402,25 @@ func (t *Team) implicitTeamDisplayName(ctx context.Context, skipConflicts bool)
 			isFullyResolved = false
 		case keybase1.TeamInviteCategory_KEYBASE:
 			// Check to make sure we don't already have the user in the name
-			iname := string(invite.Name)
```
Comment (heronhaye) [inline go/teams/teams.go]: a bit tricky... technically invite.Name for keybase-type is uv percentform, but AnnotateInvites( overwrote it with the username. I shuffled the types mainly because this was hard to follow

File: protocol/avdl/keybase1/teams.avdl [hunk 6]
```diff
@@ -619,26 +618,41 @@ protocol teams {
   @typedef("string")
   record TeamInviteName {}
 
+  // For Keybase, Seitan, Invitelink, etc. invites, we want to show a different
+  // invite name than what is in the sigchain (which may be an encoded
+  // userversion or a messagepack blob, e.g.)
+  @typedef("string")
+  record TeamInviteDisplayName {}
```
Comment (heronhaye) [inline protocol/avdl/keybase1/teams.avdl]: Clear separation of concerns from TeamInviteName which is a implementation detail more or less.

e.g. Display name will be formatted url, or max@twitter, etc whereas the name itself may be messagepacked pkey, uv percent form, whatever.

File: protocol/avdl/keybase1/teams.avdl [hunk 7]
```diff
@@ -619,26 +618,41 @@ protocol teams {
...
   }
 
   record AnnotatedTeamInvite {
-    TeamRole role;
-    TeamInviteID id;
-    TeamInviteType type;
-    TeamInviteName name;
-    UserVersion uv;
-    UserVersion inviter;
+    TeamInvite invite;
```
Comment (heronhaye) [inline protocol/avdl/keybase1/teams.avdl]: Embedded TeamInvite here so we don't accidentally drop new fields if we add them to TeamInvite

File: protocol/avdl/keybase1/teams.avdl [hunk 8]
```diff
@@ -619,26 +618,41 @@ protocol teams {
...
+
+    TeamInviteDisplayName displayName;
     string inviterUsername;
+    UserVersion inviteeUv;
```
Comment (heronhaye) [inline protocol/avdl/keybase1/teams.avdl]: Just renamed.

File: go/teams/list.go [hunk 9]
```diff
@@ -701,16 +701,11 @@ func parseInvitesNoAnnotate(ctx context.Context, g *libkb.GlobalContext, team *T
 				Implicit:       nil,
 				Status:         keybase1.TeamMemberStatus_ACTIVE,
 			})
-		} else if category == keybase1.TeamInviteCategory_SEITAN {
-			// no-op - do not parse seitans. We shouldn't even
-			// see them - they should all be stubbed out.
-		} else {
+		case keybase1.TeamInviteCategory_SEITAN, keybase1.TeamInviteCategory_INVITELINK:
+			// Don't show seitans or links in this view.
```
Comment (AMarcedone) [inline go/teams/list.go]: Which view? Maybe make the function name more explicit so it is clear what this is used for.

File: go/teams/teams.go (hunk 5 above)
Comment (AMarcedone) [reply go/teams/teams.go]: Don't understand this comment. The invite is not AnnotatedInvite here....

Comment (heronhaye) [reply go/teams/teams.go]: Not anymore. It used to be

File: go/teams/service_helper.go [hunk 10]
```diff
@@ -154,26 +154,25 @@ func Details(ctx context.Context, g *libkb.GlobalContext, name string) (res keyb
...
+	res.Members, err = MembersDetails(mctx.Ctx(), mctx.G(), t)
 	if err != nil {
 		return res, err
 	}
 
 	tracer.Stage("invites")
-	annotatedInvites, err := AnnotateInvitesUIDMapper(ctx, g, t, &res.Members)
+	res.AnnotatedActiveInvites, err = AnnotateInvitesNoPUKless(mctx, t, &res.Members)
 	if err != nil {
 		return res, err
```
[the remark this answers is not in the document]
Comment (heronhaye) [reply go/teams/service_helper.go]: Generally if err != nil there's no contract for the other values, with few exceptions (probably should be zero exceptions).

File: go/client/cmd_team_generate_invitelink.go [hunk 11]
```diff
@@ -120,11 +125,11 @@ func (c *CmdTeamGenerateInvitelink) Run() error {
 	}
 
 	dui := c.G().UI.GetDumbOutputUI()
-	dui.Printf(`Generated link: %q.
-Users can join the team by visiting that link.
+	dui.Printf(`Generated link: %s.
+Users can join the team by visiting that link and following the instructions.
 If they already have Keybase installed, they can run 'keybase team accept-invite --token %s'.
```
[the remark this answers is not in the document]
Comment (heronhaye) [reply go/client/cmd_team_generate_invitelink.go]: It should. But it may be annoying because your shell could escape stuff

[the remark this answers is not in the document]
Comment (heronhaye) [reply go/client/cmd_team_generate_invitelink.go]: so you might need quotes

File: go/teams/service_helper.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (AMarcedone) [reply go/teams/service_helper.go]: yeah. I think it is good practice to return empty structs/nil in case of errors so that if the caller forgets to check the error it is easier to notice.

File: go/teams/list.go (hunk 9 above)
Comment (heronhaye) [reply go/teams/list.go]: Not sure what a good function name here is. Do you have a suggestion?
AFAICT this is only used in `keybase team list-memberships --all` and a corresponding JSON api function.

**Comment by heronhaye** [comment]
Marked the comments I believe resolved to be resolved.

File: go/client/team_cli_rendering.go [hunk 12]
```diff
@@ -76,44 +79,42 @@ func (c *teamMembersRenderer) outputRole(team, role string, members []keybase1.T
...
+	for _, annotatedInvite := range invites {
+		invite := annotatedInvite.Invite
 		category, err := invite.Type.C()
 		if err != nil {
 			category = keybase1.TeamInviteCategory_UNKNOWN
 		}
-		trailer := fmt.Sprintf("(* invited by %s, awaiting acceptance)", invite.InviterUsername)
+		trailer := fmt.Sprintf("(* invited by %s)", annotatedInvite.InviterUsername)
+		inviteIDTrailer := ""
+		if c.showInviteID {
```
Comment (zapu) [inline go/client/team_cli_rendering.go]: would you mind also fixing this? its setup is missing in `newTeamMembersRenderer`

File: go/client/team_cli_rendering.go [hunk 13]
```diff
@@ -76,44 +79,42 @@ func (c *teamMembersRenderer) outputRole(team, role string, members []keybase1.T
...
-				res = "<token without label>"
-			}
-		}
-	}
-	return res
-}
-
 func (c *teamMembersRenderer) outputInvites(invites map[keybase1.TeamInviteID]keybase1.AnnotatedTeamInvite) {
-	for _, invite := range invites {
+	for _, annotatedInvite := range invites {
```
Comment (zapu) [inline go/client/team_cli_rendering.go]: ordering here is random on every invocation, can we sort them first?

File: go/client/team_cli_rendering.go [hunk 14]
```diff
@@ -76,44 +79,42 @@ func (c *teamMembersRenderer) outputRole(team, role string, members []keybase1.T
...
-			if c.showInviteID {
-				// Show invite IDs for SEITAN tokens, which can be used to cancel the invite.
-				inviteIDTrailer = fmt.Sprintf(" (Invite ID: %s)", invite.Id)
-			}
-			trailer = fmt.Sprintf("(* invited via secret token by %s, awaiting acceptance)%s",
-				invite.InviterUsername, inviteIDTrailer)
+			trailer = fmt.Sprintf("(* invited via secret token by %s)%s",
+				annotatedInvite.InviterUsername, inviteIDTrailer)
+		case keybase1.TeamInviteCategory_INVITELINK:
+			trailer = fmt.Sprintf("(* invitelink created by %s)%s",
```
Comment (zapu) [inline go/client/team_cli_rendering.go]: `invitelink` seems like a codename, should it say "invite link created by" for the end user?

File: go/client/team_cli_rendering.go [hunk 15]
```diff
@@ -76,44 +79,42 @@ func (c *teamMembersRenderer) outputRole(team, role string, members []keybase1.T
...
+				annotatedInvite.InviterUsername, inviteIDTrailer)
 		}
-		fmtstring := "%s\t%s*\t%s\t%s\n"
-		fmt.Fprintf(c.tabw, fmtstring, invite.TeamName, strings.ToLower(invite.Role.String()),
-			c.formatInviteName(invite), trailer)
+
+		etime := "does not expire"
+		if invite.Etime != nil {
+			t := keybase1.FromUnixTime(*invite.Etime)
+			etime = "expiration: " + humanize.RelTime(t, time.Now(), "ago", "from now")
```
Comment (zapu) [inline go/client/team_cli_rendering.go]: maybe it could say "expired:" when it's past expiration date

**Comment by zapu** [comment]
Using longer durations there doesn't work, like `8765812h` (approx 1k years). Seems like it's failing in `time` package but I couldn't track why. Maybe we could make a custom duration parser for this.

File: go/protocol/keybase1/extras.go [hunk 16]
```diff
@@ -2680,6 +2680,26 @@ func (e TeamInviteMaxUses) IsValid() bool {
...
+	if e != nil && e.IsInfiniteUses() {
+		return "unlimited uses left"
+	}
+	var maxUses int
+	if e == nil {
+		maxUses = 1
+	} else {
+		maxUses = int(*e)
+	}
+	return fmt.Sprintf("%d/%d uses left", max(maxUses-alreadyUsed, 0), maxUses)
```
Comment (zapu) [inline go/protocol/keybase1/extras.go]: I don't think that's a good format here if someone's seeing this for the first time, but I can't think of anything better.

Maybe just `Joined: %d; Limit: %d`. It's not in our designs because this was not planned for the GUI, but I think it should be since you'll be able to make these invites from cli.

File: shared/constants/teams.tsx [hunk 17]
```diff
@@ -694,22 +694,23 @@ export const annotatedInvitesToInviteInfo = (
...
       return arr
     }
 
     let username = ''
-    const t = invite.type
-    if (t.c === RPCTypes.TeamInviteCategory.sbs) {
-      const sbs: RPCTypes.TeamInviteSocialNetwork = t.sbs
-      username = `${invite.name}@${sbs}`
+    if (invite.invite.type.c === RPCTypes.TeamInviteCategory.sbs) {
+      username = invite.displayName
```
Comment (zapu) [inline shared/constants/teams.tsx]: nice

File: shared/constants/teams.tsx [hunk 18]
```diff
@@ -694,22 +694,23 @@ export const annotatedInvitesToInviteInfo = (
...
     }
     arr.push({
-      email: invite.type.c === RPCTypes.TeamInviteCategory.email ? invite.name : '',
-      id: invite.id,
-      name: invite.type.c === RPCTypes.TeamInviteCategory.seitan ? invite.name : '',
-      phone: invite.type.c === RPCTypes.TeamInviteCategory.phone ? invite.name : '',
+      email: invite.invite.type.c === RPCTypes.TeamInviteCategory.email ? invite.displayName : '',
```
Comment (zapu) [inline shared/constants/teams.tsx]: i think this might be busted for emails

File: go/teams/list.go [hunk 19]
```diff
@@ -588,26 +567,35 @@ func AnnotateInvitesUIDMapper(ctx context.Context, g *libkb.GlobalContext, team
...
+		mctx.Debug("AnnotateInvitesNoPUKless: MapUIDsReturnMap returned: %v", err)
 	}
 
 	teamName := team.Name().String()
 	for id, invite := range invites {
-		username := namePkgs[invite.Inviter.Uid].NormalizedUsername
+		inviterUsername := namePkgs[invite.Inviter.Uid].NormalizedUsername
+
+		// default displayName; overridden by some invite types later
+		displayName := keybase1.TeamInviteDisplayName(invite.Name)
```
Comment (zapu) [inline go/teams/list.go]: can we clarify that this mostly applies to phone and email?

File: go/teams/seitan_invitelink.go [hunk 20]
```diff
@@ -100,3 +102,23 @@ func GenerateSeitanInvitelinkAcceptanceKey(sikey []byte, uid keybase1.UID, eldes
 	}
 	return generateAcceptanceKey(akeyPayload, sikey)
 }
+
+// bound from SeitanEncodedIKeyInvitelinkLength
+var invitelinkIKeyRxx = regexp.MustCompile(`/invite#i=([a-z0-9+]{16,28})`)
+
+func GenerateInvitelinkURLPrefix(mctx libkb.MetaContext) (string, error) {
```
Comment (zapu) [inline go/teams/seitan_invitelink.go]: does this have to be a public function? where else could this be used?

File: go/client/team_cli_rendering.go (hunk 15 above)
Comment (heronhaye) [reply go/client/team_cli_rendering.go]: think its fine

File: go/protocol/keybase1/extras.go (hunk 16 above)
Comment (heronhaye) [reply go/protocol/keybase1/extras.go]: /shrug i think it's ok. This is displayed in tab formatted table so I was trying to be concise

File: shared/constants/teams.tsx (hunk 18 above)
Comment (heronhaye) [reply shared/constants/teams.tsx]: LMK if there's a way I can simplify the code or make it more obvious what's going on (here or in go)

**Comment by heronhaye** [comment]
@zapu want to make a ticket for the time duration thing?

**Comment by zapu** [comment]
will do, thanks
